flow/layers/image_filter_layer.cc

```diff
@@ -16,39 +16,55 @@ void ImageFilterLayer::Preroll(PrerollContext* context,
   Layer::AutoPrerollSaveLayerState save =
       Layer::AutoPrerollSaveLayerState::Create(context);
 
-  child_paint_bounds_ = SkRect::MakeEmpty();
-  PrerollChildren(context, matrix, &child_paint_bounds_);
+  SkRect child_bounds = SkRect::MakeEmpty();
+  PrerollChildren(context, matrix, &child_bounds);
   if (filter_) {
-    const SkIRect filter_input_bounds = child_paint_bounds_.roundOut();
+    const SkIRect filter_input_bounds = child_bounds.roundOut();
     SkIRect filter_output_bounds =
         filter_->filterBounds(filter_input_bounds, SkMatrix::I(),
                               SkImageFilter::kForward_MapDirection);
-    set_paint_bounds(SkRect::Make(filter_output_bounds));
-  } else {
-    set_paint_bounds(child_paint_bounds_);
+    child_bounds = SkRect::Make(filter_output_bounds);
   }
+  set_paint_bounds(child_bounds);
 
-  TryToPrepareRasterCache(context, this, matrix);
+  if (render_count_ >= kMinimumRendersBeforeCachingFilterLayer) {
+    TryToPrepareRasterCache(context, this, matrix);
+  } else {
+    render_count_++;
+    TryToPrepareRasterCache(context, GetCacheableChild(), matrix);
+  }
 }
 
 void ImageFilterLayer::Paint(PaintContext& context) const {
   TRACE_EVENT0("flutter", "ImageFilterLayer::Paint");
   FML_DCHECK(needs_painting());
 
-  if (context.raster_cache &&
-      context.raster_cache->Draw(this, *context.leaf_nodes_canvas)) {
-    return;
+  if (context.raster_cache) {
+    if (context.raster_cache->Draw(this, *context.leaf_nodes_canvas)) {
+      return;
+    }
+    const SkMatrix& ctm = context.leaf_nodes_canvas->getTotalMatrix();
+    sk_sp<SkImageFilter> transformed_filter = filter_->makeWithLocalMatrix(ctm);
+    if (transformed_filter) {
+      SkPaint paint;
+      paint.setImageFilter(transformed_filter);
+
+      if (context.raster_cache->Draw(GetCacheableChild(),
+                                     *context.leaf_nodes_canvas, &paint)) {
+        return;
+      }
+    }
   }
 
   SkPaint paint;
   paint.setImageFilter(filter_);
 
   // Normally a save_layer is sized to the current layer bounds, but in this
   // case the bounds of the child may not be the same as the filtered version
-  // so we use the child_paint_bounds_ which were snapshotted from the
-  // Preroll on the children before we adjusted them based on the filter.
-  Layer::AutoSaveLayer save_layer =
-      Layer::AutoSaveLayer::Create(context, child_paint_bounds_, &paint);
+  // so we use the bounds of the child container which do not include any
+  // modifications that the filter might apply.
+  Layer::AutoSaveLayer save_layer = Layer::AutoSaveLayer::Create(
```

Comment:
Will the `TryToPrepareRasterCache(context, GetCacheableChild(), matrix);` be wasted in this case when there's no `transformed_filter` and the count threshold is not met?

Reply:
I think what you are saying is that if the image filter cannot absorb the local matrix then we will have cached the children for nothing. That is true. I should perhaps move the attempt to apply the local transform to the preroll phase and both use it to prevent the caching and also stash it for the Paint phase.

Reply:
For the record, this would be a very rare occurrence.

Reply:
I updated this in the latest commit to check the transform application in the Preroll phase.